sources/bloodhound/src/main.rs

```diff
@@ -0,0 +1,210 @@
+into further compliance automation.
+
+# Usage
+
+Bloodhound is ultimately intended to be used throught he Bottlerocket `apiclient`
+interface. It can be executed from sheltie with the following options:
+
+```txt
+Command line arguments for the bloodhound program
+
+Usage: bloodhound [OPTIONS]
+
+Options:
+  -c, --checks <CHECK_DIR>  Path to the directory containing checker binaries [default: /usr/libexec/bloodhound-checks]
+  -f, --format <FORMAT>     Format of the output [default: text] [possible values: text, json]
+  -l, --level <LEVEL>       The CIS benchmark compliance level to check [default: 1]
+  -o, --output <OUTPUT>     Write output to a file at given path [default: stdout]
+  -h, --help                Print help
+  ```
+*/
+
+use bloodhound::output::{JsonReportWriter, ReportWriter, TextReportWriter};
+use bloodhound::results::{CheckStatus, CheckerMetadata, CheckerResult, ReportResults};
+use clap::{Parser, ValueEnum};
+use std::collections::HashMap;
+use std::fs::{DirEntry, File};
+use std::io::{stdout, Error, Write};
+use std::os::unix::fs::PermissionsExt;
+use std::process::{Command, Output};
+use std::{fs, path::PathBuf};
+
+// Define some exit codes for error conditions
+const CHECKER_DISCOVERY_ERROR: i32 = 1;
+const REPORT_OUTPUT_ERROR: i32 = 2;
+const NO_CHECKS_RUN_ERROR: i32 = 3;
+
+#[derive(ValueEnum, Clone, Debug)]
+enum Format {
+    Text,
+    Json,
+}
+
+/// Command line arguments for the bloodhound program.
+#[derive(Parser, Debug)]
+struct Arguments {
+    /// Path to the directory containing checker binaries [default: /usr/libexec/bloodhound-checks]
+    #[arg(short, long = "checks")]
+    check_dir: Option<PathBuf>,
+    /// Format of the output
+    #[clap(short, long, value_enum, default_value_t=Format::Text)]
+    format: Format,
+    /// The CIS benchmark compliance level to check
+    #[arg(short, long, default_value_t = 1)]
+    level: u8,
+    /// Write output to a file at given path [default: stdout]
+    #[arg(short, long)]
+    output: Option<PathBuf>,
+}
+
+/// Discover all executable checker files for the given directory.
+/// By looking at the provided path, this does some basic checks and filtering
+/// to find the executable files that appear to be bloodhound checkers.
+/// It will parse the checker metadata to provide a mapping between a filesystem
+/// path and a specific check, including only those checkers that are within the
+/// request compliance level.
+fn find_checkers(check_dir: &PathBuf, level: u8) -> HashMap<String, CheckerMetadata> {
+    let entries: Vec<DirEntry> = fs::read_dir(check_dir)
+        .unwrap()
+        .filter_map(|file| file.ok())
+        .collect();
+    let mut result = HashMap::with_capacity(entries.len());
+
+    for entry in entries {
+        if let Ok(file_metadata) = fs::metadata(entry.path()) {
+            // Filter out any subdirectories
+            if file_metadata.is_dir() {
+                continue;
+            }
```

Comment:
I feel doubtful about this check.

If we, the core distro developers, own the structure of this directory, we know we won't put directories in the checkers patch. However, OOTB users may do what they please and organize their checks differently e.g., in directories; then this check will ignore their checks without even providing a heads-up saying “FYI, ignoring unexpected directory”.

Thus, I think we either traverse directories that are placed here, or we print a message saying that the directories will be ignored, and document in OOTB that for this feature, `bloodhound` expects actual binaries.